File: src/modules/algorithm/seed/algorithms/maze-solver/go/bfs-solution.go
package mazesolver

type Maze [][]int
type Point struct {
	Row int
	Col int
}

// SolveMaze finds a path from start to end in a maze using Breadth-First Search (BFS).
// Returns the path as a slice of Points, or nil if no solution exists.
func SolveMaze(maze Maze, start Point, end Point) []Point {
	rows := len(maze)
	cols := len(maze[0])

	queue := []struct {
		point Point
		path  []Point
	}{{point: start, path: []Point{start}}}

	visited := make([][]bool, rows)
	for i := range visited {
		visited[i] = make([]bool, cols)
	}
	visited[start.Row][start.Col] = true

	for len(queue) > 0 {
		curr := queue[0]
		queue = queue[1:]
		point := curr.point
		path := curr.path

		if point == end {
			return path
		}

		// Explore in all four directions
		directions := []Point{{Row: -1, Col: 0}, {Row: 1, Col: 0}, {Row: 0, Col: -1}, {Row: 0, Col: 1}} // Up, Down, Left, Right

		for _, direction := range directions {
			newRow := point.Row + direction.Row
			newCol := point.Col + direction.Col

			if newRow >= 0 && newRow < rows && newCol >= 0 && newCol < cols && maze[newRow][newCol] == 0 && !visited[newRow][newCol] {
				newPoint := Point{Row: newRow, Col: newCol}
				newPath := make([]Point, len(path))
				copy(newPath, path)
				newPath = append(newPath, newPoint)

				queue = append(queue, struct {
					point Point
					path  []Point
				}{point: newPoint, path: newPath})
				visited[newRow][newCol] = true
			}
		}
	}

	return nil // No solution found
} 

File: src/modules/algorithm/seed/algorithms/maze-solver/python/bfs-solution.py
from typing import List, Tuple, Optional
from collections import deque

Maze = List[List[int]]
Point = Tuple[int, int]


def solve_maze(maze: Maze, start: Point, end: Point) -> Optional[List[Point]]:
    """
    Solves a maze using Breadth-First Search (BFS).

    Args:
        maze: The maze represented as a 2D list.
        start: The starting point in the maze (row, col).
        end: The ending point in the maze (row, col).

    Returns:
        The path from start to end as a list of (row, col) coordinates, or None if no solution exists.
    """
    rows, cols = len(maze), len(maze[0])
    queue = deque([(start, [start])])  # Queue of (current point, path so far)
    visited = set()  # Set of visited points

    while queue:
        (row, col), path = queue.popleft()

        if (row, col) == end:
            return path

        visited.add((row, col))

        # Possible movements (up, down, left, right)
        movements = [(0, 1), (0, -1), (1, 0), (-1, 0)]

        for dr, dc in movements:
            new_row, new_col = row + dr, col + dc

            # Check boundaries and wall
            if (0 <= new_row < rows and 0 <= new_col < cols and
                    maze[new_row][new_col] == 0 and (new_row, new_col) not in visited):
                new_point = (new_row, new_col)
                new_path = path + [new_point]
                queue.append((new_point, new_path))

    return None  # No path found 